swaybar/input.c

```diff
@@ -224,15 +223,14 @@ static void workspace_next(struct swaybar *bar, struct swaybar_output *output,
 static void process_discrete_scroll(struct swaybar_seat *seat,
 		struct swaybar_output *output, struct swaybar_pointer *pointer,
 		uint32_t axis, wl_fixed_t value) {
-	// If there is a button press binding, execute it, skip default behavior,
-	// and check button release bindings
 	uint32_t button = wl_axis_to_button(axis, value);
-	if (check_bindings(seat->bar, button, WL_POINTER_BUTTON_STATE_PRESSED)) {
-		check_bindings(seat->bar, button, WL_POINTER_BUTTON_STATE_RELEASED);
+	if (process_hotspots(output, pointer->x, pointer->y, button)) {
```

Comment:
Should this path now also `check_bindings(seat->bar, button, WL_POINTER_BUTTON_STATE_RELEASED);`, as you see a number of the others later in the routine?

Reply:
Mmmm, actually as far as I can see there isn't anything wrong with this press/release code on `process_discrete_scroll` you commented on, since for scroll, the press/release come immediately after one another, and currently the hotspot callback do anything for the release case.

However, it prompted me to take a deeper look into how regular (pointer, non-scroll) press/release bindsyms are handled, and I found that actually, this PR was not handling *those* release events correctly.

Since hotspots (workspaces, status bar, tray, etc.) now handle pressed events, it makes sense that they also handle released events (even if they actually do nothing), in order to "block" a `bindsym --release buttonX` which is set in the bar from triggering. Otherwise weird ensues because if you simply press/release on a hotspot without moving the mouse, you'd just see the `bindsym --release buttonX` but not a `bindsym buttonX`. This is also how i3 handles this case, and **I added a commit to this PR fixing this**.

For example, suppose we define a `bindsym button1 exec something` and a `bindsym --release button1 exec somethingelse` on the bar, and then we left click on a tray icon. Then the behavior is:

|                | Tray event | Cmd for bindsym button1 | Cmd for bindsym --release button1 |
|----------------|------------|-------------------------|-----------------------------------|
| Latest Sway    | No         | Yes                     | Yes                               |
| This PR before | Yes        | No                      | Yes                               |
| Latest i3      | Yes        | No                      | No                                |
| This PR now    | Yes        | No                      | No                                |

I also tested the various combinations (press and release on bar, press and release on hotspot, press on bar and release on hotspot, press on hotspot and release on bar) and it matches how i3 handles those.